### PR TITLE
SNOW-1801330: Map numpy functions in apply to builtin snowflake functions

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1801330

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [ ] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://github.com/snowflakedb/snowpark-python/blob/main/CONTRIBUTING.md#thread-safe-development)

3. Please describe how your code solves the related issue.

   Map numpy functions to builtin snowpark function if exists. In POC this provided 5x speedup.

### CODE REVIEW

File: tests/integ/modin/frame/test_applymap.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def test_applymap_numpy(func):
     native_df = native_pd.DataFrame(data)
     snow_df = pd.DataFrame(data)
 
-    with SqlCounter(query_count=7, udf_count=1):
+    with SqlCounter(query_count=1):
```
Comment (sfc-gh-jkew) [inline tests/integ/modin/frame/test_applymap.py:110]: This is great

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 2]
```diff
@@ -95,22 +71,66 @@
...
+    sp_func.cosh,
+    sp_func.tanh,
+    sp_func.ceil,
+    sp_func.floor,
+    sp_func.trunc,
+    sp_func.sqrt,
+    sp_func.snowflake_cortex_summarize,
+}
+
+NUMPY_FUNCTION_TO_SNOWFLAKE_FUNCTION = {
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:92]: The `numpy_to_pandas` file should changed as well, right? The numpy ufunc function mapping should be in one place.

**Review by sfc-gh-jkew — CHANGES_REQUESTED** [review]

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py (hunk 2 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:92]: They are coming from different sources. Let me see if it's possible to consolidate this into one.

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:92]: one thing we can do is change entries in numpy_to_pandas.numpy_to_pandas_universal_func_map something like this
`"exp": lambda obj, inputs: obj.apply(sp_func.exp)` -> `"exp": lambda obj, inputs: obj.apply(np.exp)` 
And then numpy to snowpark function mapping will only be in `snowflake_query_compiler.applymap`
Does it make sense or were you thinking something else? @sfc-gh-jkew

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -8755,6 +8756,12 @@ def applymap(
                     f"Snowpark pandas applymap API doesn't yet support Snowpark Python function `{func.__name__}` with args = '{args}'."
                 )
             return self._apply_snowpark_python_function_to_columns(func, kwargs)
+
+        # Check if the function is a known numpy function that can be translated to Snowflake function.
+        sf_func = NUMPY_FUNCTION_TO_SNOWFLAKE_FUNCTION.get(func)
+        if sf_func is not None:
+            return self._apply_snowpark_python_function_to_columns(sf_func, kwargs)
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:8763]: Any idea why the code coverage is complaining about this line even though it's clear that there is a test covering it? The query_count is even reduced, which means that we must have entered into this branch!

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:8763]: that is strange. need to debug this.

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 4]
```diff
@@ -95,22 +71,66 @@
...
+    np.reciprocal: lambda col: 1 / col,
+    np.exp: sp_func.exp,
+    np.exp2: lambda col: sp_func.pow(2, col),
+    np.expm1: lambda col: sp_func.exp(col) - 1,
+    np.log: sp_func.ln,
+    np.log2: sp_func._log2,
+    np.log10: sp_func._log10,
+    np.log1p: lambda col: sp_func.ln(col + 1),
+    # Aggregate functions translate to identity functions when applied element wise
+    np.sum: lambda col: col,
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:110]: Is this right? `df.apply(np.sum)` would translate to `df.apply(lambda col: col)`

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py (hunk 2 above)
Comment (sfc-gh-jkew) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:92]: That might add an unnecessary level of indirection in some cases.  
   `df.apply(np.cos)` is the same as `np.cos(df)`and ultimately it should all map to `sp_func.cos`

The np.<function> mapping can be occasionally different than the df.apply(<function>) mapping, but for simple unary functions it's probably the same. We might be able to refactor this to treat unary functions differently, but that might be to confusing to the reader.

Maybe the best idea is to move `NUMPY_FUNCTION_TO_SNOWFLAKE_FUNCTION` into `numpy_to_pandas.py` - at least then we have all numpy compat stuff in the same file.

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py (hunk 4 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:110]: Yeah when applied element-wise aggregate functions are pretty much no-op. 
Screenshot 2024-12-13 at 11 37 04 AM

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:110]: Just to clarify this is applicable to series.apply, series.map, df.applymap and df.map (which all work elementwise)

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:110]: df.apply is not impacted by this PR.

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py (hunk 2 above)
Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:92]: Moved `NUMPY_FUNCTION_TO_SNOWFLAKE_FUNCTION` into `numpy_to_pandas.py`

**Review by sfc-gh-jkew — APPROVED** [review]

File: src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py [hunk 5]
```diff
@@ -26,32 +26,7 @@
...
-    tan,
-    sinh,
-    cosh,
-    tanh,
-    ceil,
-    floor,
-    trunc,
-    sqrt,
-)
+from snowflake.snowpark import functions as sp_func
```
Comment (sfc-gh-jjiao) [inline src/snowflake/snowpark/modin/plugin/_internal/apply_utils.py:29]: Good refactor, thanks

File: tests/integ/modin/frame/test_applymap.py (hunk 1 above)
Comment (sfc-gh-jjiao) [reply tests/integ/modin/frame/test_applymap.py:110]: qq: what change has reduce the query_count from 7 to 1 and eliminate the udf call?

**Review by sfc-gh-jjiao — APPROVED** [review]

**Pull request merged**